client/src/periodUtils.js

```diff
@@ -42,16 +42,28 @@ export const PERIOD_FACTORIES = {
       .subtract(2 * offset, "weeks")
       .endOf("week")
   }),
-  [RECURRENCE_TYPE.SEMIMONTHLY]: (date, offset) => ({
-    start:
-      date.date() < 15
-        ? date.clone().subtract(offset, "months").startOf("month")
-        : date.clone().subtract(offset, "months").startOf("month").date(15),
-    end:
-      date.date() < 15
-        ? date.clone().subtract(offset, "months").startOf("month").date(14)
-        : date.clone().subtract(offset, "months").endOf("month")
-  }),
+  [RECURRENCE_TYPE.SEMIMONTHLY]: (date, offset) => {
```

Comment:
I do not think proposed logic is correct, I suggest something along
```
    const dateIsMidMonth = (date.date() > 14 ? 1 : 0)
    const periodIsMidMonth = Math.abs((dateIsMidMonth - offset ) % 2) !== 0
    const periodMonth = date.clone().subtract(Math.ceil((dateIsMidMonth + offset + (offset > 0 ? 1 : 0) )/2 ) , "months")
            
    return {
    	periodIsMidMonth ? 
	    {
	    start: monthDate.clone().startOf("month"),
	    end: monthDate.clone().date(14).endOf(day)
	    }
	:
	    {
	    start: monthDate.clone().date(15),
	    end: monthDate.clone().endOf("month").endOf(day)
	    } 
    }
```

Reply:
Do you have a concrete case to show that the logic is incorrect? Also, in your proposed solution, what are you trying to achive with these lines?:
`const dateIsMidMonth = (date.date() > 14 ? 1 : 0)`
`const periodIsMidMonth = Math.abs((dateIsMidMonth - offset ) % 2) !== 0`
It doesn't makes sense to talk about mid of month when there two halves. Is second half of the month is mid Month?

Reply:
Hmm, i just found that it does result in wrong periods if current date is 16th of a month and that month is after a 31-day month. This is complicated :) I will simplify with a correct solution.

Reply:
dateIsMidMonth <=> Date is in second period of month
periodIsMidMonth <=> period is the second period of the month
feel free to rename. I have not tested the logic (also did a small fix), but it seems more readable at least to me.

In your case (again, just by reading only):

date = 1 April, offset = 0
base date = 1 April
isFirstHalfOfTheMonth = true
baseMonthStart = april
start = 16 April
end = 31 April

Reply:
What do you think about this ? 
```
    const isCurDateInFirstHalf = date.date() < 15

    const isPeriodFirstHalf =
      (isCurDateInFirstHalf && offset % 2 === 0) ||
      (!isCurDateInFirstHalf && offset % 2 === 1)

    // Don't subtract 15 days, fails when Febuary being 28-29 days and some months 31 days
    // instead subtract months(dividing offset by 2) to round up edge cases (moment takes care of it)
    // if offset is odd, divide by ceiling number and add 15 days
    const baseDate =
      offset % 2 === 0
        ? date.clone().subtract(offset / 2, "months") // simple case when even offset
        : isCurDateInFirstHalf // if it is odd, we need to adjust according to isCurDateInFirstHalf
          ? date.clone().subtract(Math.ceil(offset / 2), "months")
          : date.clone().subtract(Math.floor(offset / 2), "months")

    const baseMonthStart = baseDate.clone().startOf("month")

    return isPeriodFirstHalf
      ? {
        start: baseMonthStart,
        end: baseMonthStart.clone().endOf("day").add(13, "days") // end of day 14
      }
      : {
        start: baseMonthStart.clone().add(14, "days"), // start of day 15
        end: baseMonthStart.clone().endOf("month")
      }
  }
```

Reply:
I have made a correct working version i think, needs testing

Reply:
Way too complicated, look at https://github.com/NCI-Agency/anet/blob/GH-3231-pending-assessments-email-notification/src/main/java/mil/dds/anet/threads/PendingAssessmentsNotificationWorker.java#L319-L367 for some inspiration.

Reply:
Are you dealing with offset there ?

Reply:
Because that is what makes it complicated. Let me explain the problem:
- You can't add / subtract 15 days. As the months can be 28-29-30-31 days.
- We have to do the addition/subtraction in months
- We let momentjs figure out how the months play out. We just need to figure out how many months we need to subtract corresponding to `offset` value (`offset` / 2 in even number case since semimonth, odd case is complicated, also there are negative and positive values as well)

Reply:
Just look at the Java code: https://github.com/NCI-Agency/anet/blob/GH-3231-pending-assessments-email-notification/src/main/java/mil/dds/anet/threads/PendingAssessmentsNotificationWorker.java#L317-L368
It does the whole calculation in months at the beginning of the month, then may add 15 days.

Reply:
I did look at the code. You are calculating the previous period right? So you will check if there are assessments on that previous period. What i am trying to do is different. I am calculating every period given offset for `Periodic Assessments Table`. Try to answer this question, what would you do if you were not calculating previous period but nth previous period? That is where the complications arise.

Reply:
1 semimonth before a First Half (1-14) period is previous month, but 1 semimonth before a Second Half (15-endOf) is same month. 
1 semimonth after a First Half is same month, but 1 semimonth after a Second Half is next month :) and so on...

Reply:
```
    final ZonedDateTime zonefulNow = now.atZone(DaoUtils.getDefaultZoneId());
    […]
    final ZonedDateTime bom =
        zonefulNow.with(TemporalAdjusters.firstDayOfMonth()).truncatedTo(ChronoUnit.DAYS);
    […]
    final ZonedDateTime result;
    switch (recurrence) {
      […]
      case RECURRENCE_SEMIMONTHLY: // two per month: [1 - 14] and [15 - end-of-month]
        final int startDayOfSecondPeriod = 15;
        result = zonefulNow.get(ChronoField.DAY_OF_MONTH) >= startDayOfSecondPeriod ? bom
            : bom.minus(1, ChronoUnit.MONTHS).plus(startDayOfSecondPeriod, ChronoUnit.DAYS);
        break;
    […]
    }
    return result.toInstant();
```
What more can I say?

Reply:
Except that `.plus(startDayOfSecondPeriod, ChronoUnit.DAYS)` should probably be `.plus(14, ChronoUnit.DAYS)` as `1 + 15 == 16` :laughing:

Reply:
Changed that bit of code to:
```
       case RECURRENCE_SEMIMONTHLY: // two per month: [1 - 14] and [15 - end-of-month]
-        final int startDayOfSecondPeriod = 15;
-        result = zonefulNow.get(ChronoField.DAY_OF_MONTH) >= startDayOfSecondPeriod ? bom
-            : bom.minus(1, ChronoUnit.MONTHS).plus(startDayOfSecondPeriod, ChronoUnit.DAYS);
+        final int daysInFirstPeriod = 14;
+        result = zonefulNow.get(ChronoField.DAY_OF_MONTH) > daysInFirstPeriod ? bom
+            : bom.minus(1, ChronoUnit.MONTHS).plus(daysInFirstPeriod, ChronoUnit.DAYS);
         break;
```

Reply:
@gjvoosten Do you read my comments? :confused: If the problem was calculating the last period for checking if there are pending assessments, yes your solution is helpful, otherwise, the general question of calculating nth previous or nth next period {start, end} from current date is not simple as your java code. There are edge cases involved

Reply:
Well, the Java worker does not need the end date (JavaScript side does, indeed), but calculating the end date should also not be that complicated. Basically the end date is the previous day of the start date of the next period.

Reply:
Anyway, what about:
```
  [RECURRENCE_TYPE.SEMIMONTHLY]: (date, offset) => ({
    start:
      date.date() < 15
        ? date.clone().subtract(offset, "months").startOf("month").date(15).startOf("day"),
        : date.clone().subtract(offset - 1, "months").startOf("month")
    end:
      date.date() < 15
        ? date.clone().subtract(offset, "months").endOf("month")
        : date.clone().subtract(offset - 1, "months").startOf("month").date(14).endOf("day")
  }),
```
?

Reply:
`offset` is in terms of semi-months

Reply:
Ah, yes. Hmm…

Reply:
The code can definitely be made more readable, but i think i captured the logic just fine. Math.ceil and Math.floor is basically if else conditions for First Half and Second Half calculations, those math functions make it harder to understand as negative numbers become confusing with those functions.
Logic is this: (skip even offsets, they are trivial, negative offset means future period)
`case (first-half and offset positive ) : offset (2n+1) => n.5 month  => n+1 month subtraction`
`case (first-half and offset negative ) :offset -(2n+1)  => -n.5 month  => -n month `
`case (second-half and offset positive ) : offset (2n+1) => n.5 month  => n month `
`case (second-half and offset negative ) :offset -(2n+1)  => -n.5 month  => -(n+1) month `